### PR TITLE
ci: Fix model change detection script and registry parsing logic

### PR DESCRIPTION
**Pull request by tadani3**
This pull request significantly improves the logic for detecting which models are affected by changes in the codebase, particularly for changes to component and task files. Instead of always triggering a full rebuild/test (`run_all`) when these files change, the script now traces their impact via the import graph to determine exactly which models are affected. The registry parsing is also enhanced to support a new declarative registration pattern. Comprehensive tests have been added and updated to validate these behaviors.

**Improvements to affected model detection:**

* Component and task files (under `src/mobius/components/` and `src/mobius/tasks/`) are now classified as `"traceable"` instead of `"shared_infra"`, and their impact is determined by tracing the import graph to find affected models, rather than always triggering `run_all`. (`scripts/detect_affected_models.py`, `scripts/detect_affected_models_test.py`) [[1]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76L51-R54) [[2]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76L60-R63) [[3]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76L81-R88) [[4]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89L45-R49) [[5]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89L176-R189) [[6]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89R271-R370)
* If a traceable file is deleted, the script conservatively triggers `run_all`. (`scripts/detect_affected_models.py`, `scripts/detect_affected_models_test.py`) [[1]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76R462-R472) [[2]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89R271-R370)

**Enhancements to registry parsing:**

* The registry mapping logic now supports a fourth pattern: declarative dictionary-based registration (e.g., `_REGISTRATIONS = {"name": ModelRegistration(ClassName, ...)}`), ensuring all registered models are correctly discovered. (`scripts/detect_affected_models.py`) [[1]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76L239-R248) [[2]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76R270-R284) [[3]](diffhunk://#diff-a2056def73ca73f9afeffb51ece97153e1ff1646daf027a79d825724dca26f76R362-R377)

**Test coverage and validation:**

* Tests have been updated and expanded to cover the new traceable logic, including integration tests for the import graph tracing and various edge cases (e.g., combined changes, subset/superset relationships, deleted files). (`scripts/detect_affected_models_test.py`) [[1]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89L176-R189) [[2]](diffhunk://#diff-a7ac8a76fe32fc8fa571202b42fa108af03ac87b1f81f8fc65cbb93df8f04f89R271-R370)
* Test assertions have been fixed and clarified, including registry expectations and affected model checks. (`scripts/detect_affected_models_test.py`)

These changes make the affected model detection more accurate and efficient, reducing unnecessary full runs and providing better granularity for CI/CD workflows.

### CODE REVIEW

File: scripts/detect_affected_models.py [hunk 1]
```diff
@@ -236,10 +241,11 @@ def _build_class_to_source_module() -> dict[str, str]:
 def _build_registry_class_to_types() -> dict[str, list[str]]:
     """Parse _registry.py to map class names to registered model_types.
 
-    Handles three patterns:
+    Handles four patterns:
     1. Direct: reg.register("name", ClassName)
     2. For-loop: for name in (...): reg.register(name, ClassName)
```
Comment (justinchuby) [inline scripts/detect_affected_models.py]: I think these are obsolete

File: scripts/detect_affected_models.py [hunk 2]
```diff
@@ -236,10 +241,11 @@ def _build_class_to_source_module() -> dict[str, str]:
...
     3. Dict-loop: for name, cls in {...}.items(): reg.register(name, cls)
+    4. Declarative dict: _REGISTRATIONS = {"name": ModelRegistration(ClassName, ...)}
```
Comment (justinchuby) [inline scripts/detect_affected_models.py]: This is the new pattern we migrated to

**Review by justinchuby — APPROVED** [review]

File: scripts/detect_affected_models.py [hunk 3]
```diff
@@ -451,6 +496,22 @@ def detect_affected_models(
...
+        # → mobius.components._attention
+        rel = normalized[len("src/") :]
+        module_name = rel[:-3].replace("/", ".")  # strip .py, dots
+        if not module_name:
+            continue
+
+        dependents = _find_reverse_dependents(module_name, import_graph)
+        for dep_module in dependents:
+            if dep_module in registry_map:
+                affected.update(registry_map[dep_module])
```
[the remark this answers is not in the document]
Comment (tadani3) [reply scripts/detect_affected_models.py:426]: Fixed, moved tasks back to SHARED_INFRA for the time being

File: scripts/detect_affected_models.py [hunk 4]
```diff
@@ -451,6 +496,22 @@ def detect_affected_models(
...
                 affected.update(registry_map[dep_module])
 
+    # Process traceable files (components, tasks): find which models
+    # transitively import them, then map to registered model_types.
+    for path in traceable_files:
+        normalized = path.replace("\\", "/")
+        # Convert path to module name: src/mobius/components/_attention.py
+        # → mobius.components._attention
+        rel = normalized[len("src/") :]
+        module_name = rel[:-3].replace("/", ".")  # strip .py, dots
```
[the remark this answers is not in the document]
Comment (tadani3) [reply scripts/detect_affected_models.py]: Resolved, fixed in a slightly different way. Opus 4.6 caught another possible failure point.

File: scripts/detect_affected_models_test.py [hunk 5]
```diff
@@ -173,13 +174,19 @@ def test_reverse_dependents_no_self(self):
...
+        assert len(result["affected"]) > 0
 
-    def test_task_change_triggers_run_all(self):
+    def test_task_change_traces_affected_models(self):
+        """A task change traces through the import graph to find affected models."""
         result = detect_affected_models(["src/mobius/tasks/_causal_lm.py"])
-        assert result["run_all"] is True
+        assert result["run_all"] is False
+        # _causal_lm.py is imported by task infrastructure — may affect models
+        # The key point: it does NOT trigger run_all
```
[the remark this answers is not in the document]
Comment (tadani3) [reply scripts/detect_affected_models_test.py]: Already handled — the test explicitly asserts [run_all is True](vscode-file://vscode-app/c:/Users/tommasoadani/AppData/Local/Programs/Microsoft%20VS%20Code/560a9dba96/resources/app/out/vs/code/electron-browser/workbench/workbench.html). No changes needed.

File: scripts/detect_affected_models.py [hunk 6]
```diff
@@ -78,9 +81,11 @@ def classify_file(path: str) -> str:
     # Shared infrastructure patterns
     if normalized in _SHARED_INFRA_PATTERNS:
         return "shared_infra"
-    for prefix in _SHARED_INFRA_PREFIXES:
+
+    # Traceable infrastructure (components, tasks) — traced via import graph
+    for prefix in _TRACEABLE_PREFIXES:
         if normalized.startswith(prefix):
-            return "shared_infra"
+            return "traceable"
```
[the remark this answers is not in the document]
Comment (tadani3) [reply scripts/detect_affected_models.py]: Resolved

File: scripts/detect_affected_models.py (hunk 1 above)
Comment (tadani3) [reply scripts/detect_affected_models.py]: Resolved. Eliminated reference to obsolete patterns to clean up the script.

File: scripts/detect_affected_models.py (hunk 2 above)
Comment (tadani3) [reply scripts/detect_affected_models.py]: Resolved. The script now handles this pattern only.
